### PR TITLE
Prep for MS Teams plugin API

### PR DESCRIPTION
**Pull request by wiggin77**
#### Summary
Preparation tasks for Shared Channels to support plugin API for MS Teams.

- add plugid to remoteclusters table and store
- add lastpostcreateat and lastpostcreateid to sharedchannelremotes table and store
- sync new posts before updated posts to ensure post order in MS Teams
- don't sync history by default

Developer tested against 9.3 server to ensure Shared Channels can still sync with older servers.

#### Ticket Link
https://mattermost.atlassian.net/browse/MM-55965

#### Release Note
```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
Any reason you are waiting for MySQL 8? We are still allowing MySQL 5 to start with the server. I'd suggest to revert the MySQL changes.

File: server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+        AND column_name = 'LastPostUpdateID'
+    ),
+        IF(
+            EXISTS(
+                SELECT 1 FROM INFORMATION_SCHEMA.COLUMNS
+                WHERE table_name = 'SharedChannelRemotes'
+                AND table_schema = DATABASE()
+                AND column_name = 'LastPostID'
+        ), 
+        'ALTER TABLE SharedChannelRemotes RENAME COLUMN LastPostID TO LastPostUpdateID;',
```
Comment (agnivade) [inline server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql:64]: From the docs [here](https://dev.mysql.com/doc/refman/5.7/en/innodb-online-ddl-operations.html#online-ddl-column-syntax-notes), I'd suggest: `ALTER TABLE tbl CHANGE old_col_name new_col_name data_type, ALGORITHM=INPLACE, LOCK=NONE;`

**Review by agnivade — CHANGES_REQUESTED** [review]

File: server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+    ),
+    'ALTER TABLE SharedChannelRemotes ADD COLUMN LastPostCreateID varchar(26);',
+    'SELECT 1;'
+));
+
+PREPARE addColumnIfNotExists FROM @preparedStatement;
+EXECUTE addColumnIfNotExists;
+DEALLOCATE PREPARE addColumnIfNotExists;
+
+-- rename `lastpostid` column only if it exists and the new name doesn't exist.
```
Comment (lieut-data) [inline server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql:48]: Is this a breaking schema change, in that peers in the HA cluster that upgrade incrementally will immediately break?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Early feedback, just to ask more details about what seems like a breaking schema change.

File: server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql (hunk 2 above)
Comment (agnivade) [reply server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql:48]: Yes indeed. Good point. I think I had this in my doc somewhere, but yeah we cannot rename the column atomically any more. Our schema migration guidelines guarantee that old nodes in an HA cluster will still work correctly as the new node gets migrated.

IF we want to rename this, we need to:
1. Create the new column with the new name.
2. Start a batch migration worker which will slowly migrate values from the old column to the new, and at the same time, do a double write from within the code.
3. In the next ESR, drop the old column, and start using only the new column.

Long story short, it's a long winded process now. If it's not too much of a problem, I might suggest to not do the rename altogether.

Comment (wiggin77) [reply server/channels/db/migrations/mysql/000117_msteams_shared_channels.up.sql]: Rename of column has been reverted.

**Review by agnivade — APPROVED** [review]
Migrations look good. :+1:

**Review by lieut-data — APPROVED** [review]
This is really cool, thank you for exposing me to the internals, @wiggin77! A few comments inline, but nothing blocking.

Also, out of curiosity, what's the `prefix` for in the various table queries?

File: server/channels/store/sqlstore/shared_channel_store.go [hunk 3]
```diff
@@ -398,8 +403,10 @@ func sharedChannelRemoteFields(prefix string) []string {
 		prefix + "IsInviteAccepted",
 		prefix + "IsInviteConfirmed",
 		prefix + "RemoteId",
+		prefix + "LastPostCreateAt",
+		"COALESCE(" + prefix + "LastPostCreateID,'') AS LastPostCreateID",
 		prefix + "LastPostUpdateAt",
-		prefix + "LastPostId",
+		"COALESCE(" + prefix + "LastPostId,'') AS LastPostUpdateID",
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: `LastPostCreateID` is a new field that is not being back-filled.  At some point in my dev testing I ended up with a null LastPostID.  It is a null-able field, because it was added in a migration after table creation and not back-filled.

File: server/platform/services/remotecluster/ping.go [hunk 4]
```diff
@@ -106,6 +106,7 @@ func (rcs *Service) pingRemote(rc *model.RemoteCluster) error {
 	if err != nil {
 		return err
 	}
+	rc.LastPingAt = model.GetMillis()
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/platform/services/remotecluster/ping.go:109]: It would be slightly inaccurate if we wait until after the DB call to set this.  Probably not a big deal.

File: server/platform/services/sharedchannel/sync_send_remote.go [hunk 5]
```diff
@@ -186,43 +193,67 @@ func (scs *Service) fetchUsersForSync(sd *syncData) error {
...
+	sd.resultRepeat = count >= MaxPostsPerSync
+
+	return nil
+}
 
+func appendPosts(dest []*model.Post, posts []*model.Post, postStore store.PostStore, timestamp int64) []*model.Post {
 	// Append the posts individually, checking for root posts that might appear later in the list.
 	// This is due to the UpdateAt collision handling algorithm where the order of posts is not based
 	// on UpdateAt or CreateAt when the posts have the same UpdateAt value. Here we are guarding
 	// against a root post with the same UpdateAt (and probably the same CreateAt) appearing later
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/platform/services/sharedchannel/sync_send_remote.go:244]: We can get a root post and a child post with the same CreateAt and UpdateAt.  The order they will be retrieved is not guaranteed, and it's possible the child appears before the root, causing the insert of the child to fail on the remote.  Sending new posts first solves a different problem, but unfortunately not this one.

File: server/channels/store/sqlstore/post_store.go [hunk 6]
```diff
@@ -1395,10 +1396,17 @@ func (s *SqlPostStore) GetPostsSinceForSync(options model.GetPostsSinceForSyncOp
 	query := s.getQueryBuilder().
 		Select("*").
 		From("Posts").
-		Where(sq.Or{sq.Gt{"Posts.UpdateAt": cursor.LastPostUpdateAt}, sq.And{sq.Eq{"Posts.UpdateAt": cursor.LastPostUpdateAt}, sq.Gt{"Posts.Id": cursor.LastPostId}}}).
 		OrderBy("Posts.UpdateAt", "Id").
 		Limit(uint64(limit))
 
+	if options.SinceCreateAt {
+		query = query.Where(sq.Or{sq.Gt{"Posts.CreateAt": cursor.LastPostCreateAt},
+			sq.And{sq.Eq{"Posts.CreateAt": cursor.LastPostCreateAt}, sq.Gt{"Posts.Id": cursor.LastPostCreateID}}})
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/post_store.go]: This was missing some commas.  Fixed.

File: server/platform/services/sharedchannel/sync_send_remote.go [hunk 7]
```diff
@@ -41,12 +42,13 @@ type syncData struct {
...
-		users:            make(map[string]*model.User),
-		profileImages:    make(map[string]*model.User),
-		resultNextCursor: model.GetPostsSinceForSyncCursor{LastPostUpdateAt: scr.LastPostUpdateAt, LastPostId: scr.LastPostId},
+		task:          task,
+		rc:            rc,
+		scr:           scr,
+		users:         make(map[string]*model.User),
+		profileImages: make(map[string]*model.User),
+		resultNextCursor: model.GetPostsSinceForSyncCursor{LastPostUpdateAt: scr.LastPostUpdateAt, LastPostUpdateID: scr.LastPostUpdateID,
+			LastPostCreateAt: scr.LastPostCreateAt, LastPostCreateID: scr.LastPostCreateID},
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/platform/services/sharedchannel/sync_send_remote.go]: Missing comma.  Fixed.

File: server/channels/store/sqlstore/shared_channel_store.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (lieut-data) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: Ah, got it, so maybe the question is: why is this not being backfilled, when we're doing it for the other new columns?
```
ALTER TABLE RemoteClusters ADD COLUMN IF NOT EXISTS PluginID VARCHAR(190) NOT NULL DEFAULT '';

ALTER TABLE SharedChannelRemotes ADD COLUMN IF NOT EXISTS LastPostCreateAt bigint NOT NULL DEFAULT 0;

ALTER TABLE SharedChannelRemotes ADD COLUMN IF NOT EXISTS LastPostCreateID VARCHAR(26);
```

[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: @lieut-data the `RemoteClusters` table is always going to be small, however the `SharedChannelRemotes` table can get large, especially when we start sharing all channels.  That table gets a record per per per remote.  My understanding is that adding a column with no-null and default in older MySQL requires a re-write of the table.

[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: That's true, and a mistake on my part.  Now I'm second guessing if I should be adding the default to the other columns. The migration guide from server team seems to suggest adding nullable columns is fine, but doesn't explicitly state non-null with defaults is not.  Looking at MySQL documentation, there is a table re-write for older versions.  Despite EOL on MySQL versions, I don't think we can just will customers to upgrade on our timeline.  I'll use coalesce for now and not create migration headaches for folks straggling on DB upgrades.

[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: @lieut-data looking at other migrations, I do see columns added with defaults, including the reactions table.  I'm now second guessing whether I can use defaults or not.

[the remark this answers is not in the document]
Comment (agnivade) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: You can see the table here: https://dev.mysql.com/doc/refman/5.7/en/innodb-online-ddl-operations.html#online-ddl-column-operations. Adding a column rebuilds the table, but it permits concurrent DML. That's the most important part. We don't care how long it takes, as long as it does not block the table. 

We can also easily verify by adding a new column with default on the posts table and trying to make some queries and see if they return.

[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:409]: Thanks for the detail agnivade.  I'll keep that in mind for future schema changes.

File: server/channels/store/sqlstore/shared_channel_store.go [hunk 8]
```diff
@@ -532,14 +539,32 @@ func (s SqlSharedChannelStore) GetRemoteForUser(remoteId string, userId string)
...
+	if cursor.LastPostUpdateAt > 0 || cursor.LastPostUpdateID != "" {
+		builder = builder.Set("LastPostUpdateAt", cursor.LastPostUpdateAt)
+		builder = builder.Set("LastPostId", cursor.LastPostUpdateID)
+		updateNeeded = true
+	}
+
+	if !updateNeeded {
+		// no new cursor provided.
+		return fmt.Errorf("cursor empty")
+	}
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:565]: Probably not strictly an error, but is unexpected and logged.  To get in this state a `IsCursorEmpty` check was missed somewhere, OR a data sync happened, with actual data sync'd yet the cursor came back from the store empty.  I think it's better that callers are not sending empty cursors here which would be a fault in the logic of the caller.  A silent failure to update a cursor that is expected to have data would hide a bug.

[the remark this answers is not in the document]
Comment (lieut-data) [reply server/channels/store/sqlstore/shared_channel_store.go:565]: If one wanted to reset the cursor to the beginning, would we (later) expose a discrete API to do so?

**Review by mgdelacroix — APPROVED** [review]
This is great! :+1:

File: server/channels/store/sqlstore/shared_channel_store.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/channels/store/sqlstore/shared_channel_store.go:565]: @lieut-data If someone wanted all history they can pass a cursor set to 20 years ago.  I don't think we want to use the zero value timestamp for this.

File: server/platform/services/sharedchannel/service.go [hunk 9]
```diff
@@ -23,7 +23,7 @@ const (
 	TopicChannelInvite           = "sharedchannel_invite"
 	TopicUploadCreate            = "sharedchannel_upload"
 	MaxRetries                   = 3
-	MaxPostsPerSync              = 12 // a bit more than one typical screenfull of posts
+	MaxPostsPerSync              = 50 // a bit more than 4 typical screenfulls of posts
```
[the remark this answers is not in the document]
Comment (wiggin77) [reply server/platform/services/sharedchannel/service.go:26]: @mgdelacroix  There is a bit of overhead in retrieving the posts to be sent, and the change to send new posts first, then updated posts exacerbates that.  To mitigate we are sending more posts per batch.  The original value (12) was based on the original use case where there would be severely poor connectivity and smaller batches meant less likely to fail and retries aren't as large.  The use cases now have decent connectivity.  That said, this really should be a configuration option.  I've created a [ticket](https://mattermost.atlassian.net/browse/MM-56026) for that.
